android/library/src/main/java/com/liferay/mobile/screens/ddm/form/service/APIOSubmitService.kt

```diff
@@ -0,0 +1,78 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ * <p>
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ * <p>
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.mobile.screens.ddm.form.service
+
+import com.liferay.apio.consumer.exception.ThingNotFoundException
+import com.liferay.apio.consumer.exception.ThingWithoutOperationException
+import com.liferay.apio.consumer.model.Operation
+import com.liferay.apio.consumer.model.Relation
+import com.liferay.apio.consumer.model.Thing
+import com.liferay.mobile.screens.ddl.form.util.FormConstants
+import com.liferay.mobile.screens.ddl.model.Field
+import com.liferay.mobile.screens.ddm.form.serializer.FieldValueSerializer
+import okhttp3.HttpUrl
+
+/**
+ * @author Paulo Cruz
+ */
+class APIOSubmitService : ISubmitService, BaseAPIOService() {
+
+	fun submit(formThing: Thing, currentRecordThing: Thing?, fields: MutableList<Field<*>>,
```

Comment:
Is there anything here that should be in the consumer?

Reply:
Today we have to fetch the resource again because sometimes we don't have the operations inside the thing. For example (line 48):

```
getThing(thingId, { thing ->
	thing.getOperation(operationId)?.let { operation ->
		performSubmit(thing, operation, fields, isDraft, onSuccess, onError)
	} ?: onError(ThingWithoutOperationException(thing.id, operationId))
}, onError)
```

But the goal is just to perform the operation using thingId (we already have) and the operationId (we have to request again to get it). (line 69)

```
apioConsumer.performOperation(thing.id, operation.id, fillFields = { _ ->
		mapOf(
			Pair("isDraft", isDraft),
			Pair(FormConstants.FIELD_VALUES, FieldValueSerializer.serialize(fields) { !it.isTransient })
		)
	}, onComplete = { result ->
		result.fold(onSuccess, onError)
	}
)
```

Since this thing is a FormRecord, we get the first object after a create a record and then we keep the reference and update it. Since the record returned in a create operation it doesn't contain it's self-operations. 

I guess we could abstract this steps to other devs in a future refactoring. This is the only point that I can see, @nhpatt do you have another?

Reply:
So, basically, we should return the operations after a create, right?

Reply:
And here: private fun getThing(thingId) we should support strings also, right?

Reply:
I think you can cache the operations from the profile (/o/api/doc)

Reply:
Ok, I'll add the suggestion to the apio backlog